pkg/kubelet/apis/config/v1beta1/defaults.go

```diff
@@ -40,6 +40,8 @@ var (
 	// TODO: Move these constants to k8s.io/kubelet/config/v1beta1 instead?
 	// Refer to [Node Allocatable](https://git.k8s.io/community/contributors/design-proposals/node/node-allocatable.md) doc for more information.
 	DefaultNodeAllocatableEnforcement = []string{"pods"}
+	// DefaultDNSOptionsValue includes default dns options
+	DefaultDNSOptionsValue = []string{"ndots:5"}
```

Comment:
Rather than making this default be part of the API, I think we'd be better off saying that the default values is "" and the meaning of that is defined internally.  This leaves maximum flexibility to the CRI implementation, which might be OS specific.

Reply:
To understand you correctly. Your proposition is to get rid of the default values in the kubelet api and extend the logic it in `pkg/kubelet/network/dns/dns.go` in case `defaultDNSOptions` will not be specified, am I correct?

Reply:
If we pursue this path, yes.  That said, I am increasingly thinking that small policy objects are better than configs/flags.

Reply:
So @thockin to understand you correctly you would like to have small object like:
```
dnsConfig:
  options:
    - name: ndots
      value: "2"
    - name: edns0
```
to be able to extend this (or replace as currently we can override this from /etc/resolve.conf) with more properties like nameservers or searches, correct?

Reply:
Sorry, no.  I am implying a full-fledged API object that users can create per-namespace and maybe per-cluster (or maybe vice-versa?) that sets defaults for DNS.  I know that is a significantly more involved change, but it means that we never have to debate this again.

That said, it's not clear to me exactly how that would work.  We need a design doc (KEP) to work through the options.

Or maybe I am over-blowing it, and we don't need that, which I think a KEP would help to determine.

Reply:
I understand your point of view and to be honest I would like to have more generic way which can be applied per namespace or per cluster but for now there are a lot of peoples (including our organisation) which are affected by 5 seconds dns timeout issue and can't really configure needed dns options (like timeout or ndots). DNS per pod configuration or modification for /etc/resolv.conf during preStart (which we are using btw) are a bit hacky in my opinion and we should propose something to kubernetes users to minimize this problem. Ticket for this problem https://github.com/kubernetes/kubernetes/issues/59031 is almost one year and no one propose any better solutions. Initially my proposition was related with fact that dns options are the only dns properties that are hard coded and we can't override them (for example search and servers can be overdried by /etc/resolve.conf from kubernetes node).

Reply:
You've got me ;)
Sure but who will work on this to catch 1.14? I can try to support this but I'm not kubernetes developer (I don't even know go - it was just a pair programming with google)

Reply:
Go is easy!  Kubernetes is not that hard, though adding APIs is one of the more tricky parts.

Like so many ideas, it needs an owner - someone to move the ball across the goal-line.  It can't be me, though I am happy to help shepherd it.

If you can't sign up for it, we can try to find someone else to do it from sig-network or sig-node.

The starting point is a KEP - a doc that explains what we want to do in enough detail that people can debate the relative tradeoffs.

Reply:
Can I vote for @MrHohn for this role? For the implementation part  I can work on this with small help.

Reply:
You can't generally sign up other people :)  It's up to him to decide if he has time to help implement.

Reply:
@dawidmalina Thanks for your vote! Unfortunately I won't have enough bandwidth to drive this through in the 1.14 time frame :( But would definitely be happy to help on reviewing or a portion of implementation if possible.

Reply:
@thockin, @MrHohn I can try to implement this but can you please guide me a bit. As I understood we would like to have cluster or namespace level object where we will have ability to define dns policy. Do we already have such objects? If yes then I could check the code and learn how it should be implemented? I would be very happy if you could give me some hints.

Reply:
Before you dig toooo far into code, I would write a doc (KEP in kubernetes/enhancements repo) .  Start pretty high-level and iterate.  State the problem and a minimal idea to solve.

One thing we'll have to figure is whether this is a CRD or built-in.